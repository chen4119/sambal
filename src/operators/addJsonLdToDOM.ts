import {pipe, Observable} from "rxjs";
import {map} from "rxjs/operators";
import {SambalData} from "../constants";
import {toJsonLdGraph, SCHEMA_CONTEXT} from "sambal-jsonld";

export function addJsonLdToDOM() {
    return pipe<Observable<SambalData>, Observable<SambalData>>(
        map((data) => {
            if (data.html && data.jsonld && data.jsonld.length > 0) {
                const schemaOrgJson = toJsonLdGraph(data.jsonld, SCHEMA_CONTEXT);
                if (schemaOrgJson) {
                    const $ = data.html;
                    const jsonLdBlock = $('<script type="application/ld+json"></script>').appendTo($("head"));
                    jsonLdBlock.text(JSON.stringify(schemaOrgJson));
                }
            }
            return data;
        })
    );
}