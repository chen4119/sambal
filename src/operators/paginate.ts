import {Observable} from "rxjs";

export function paginate() {
    return (observable) => new Observable(observer => {
        let prevPage = null;
        let currentItemCount = 1;
        let currentPageCount = 1;
        const subscription = observable.subscribe({
            next(value) {
                const currentPage = {
                    page: currentPageCount,
                    begin: currentItemCount,
                    end: currentItemCount + value.length - 1,
                    items: value,
                    hasNext: true
                };
                if (prevPage !== null) {
                    observer.next(prevPage);
                }
                prevPage = currentPage;
                currentItemCount = currentPage.end + 1;
                currentPageCount++;
            },
            error(err) {
                observer.error(err);
            },
            complete() {
                if (prevPage !== null) {
                    prevPage.hasNext = false;
                    observer.next(prevPage);
                }
                observer.complete();
            }
        });
        return () => {
            subscription.unsubscribe();
        }
    });
}