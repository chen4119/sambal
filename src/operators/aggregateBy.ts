import {queryData} from "../Utils";
import {from, of, pipe} from "rxjs";
import {mergeMap, map, groupBy, reduce, toArray} from "rxjs/operators";

export function aggregateBy(field: string) {
    return pipe(
        map((data) => queryData(data, field)),
        mergeMap(fieldValue => Array.isArray(fieldValue) ? from(fieldValue) : of(fieldValue)),
        groupBy(v => v),
        mergeMap(group$ =>
            group$.pipe(reduce((acc, cur) => {
                acc.count++;
                return acc;
            }, {key: group$.key, count: 0}))    
        ),
        toArray()
    );
}