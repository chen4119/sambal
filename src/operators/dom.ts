import {pipe} from "rxjs";
import {tap} from "rxjs/operators";
import {SAMBAL_INTERNAL} from "../constants";

export function dom(mapper: ($: CheerioStatic) => void) {
    return pipe(
        tap(d => {
            if (d[SAMBAL_INTERNAL] && d[SAMBAL_INTERNAL].html) {
                const $ = d[SAMBAL_INTERNAL].html;
                mapper($);
            }
        })
    );
}