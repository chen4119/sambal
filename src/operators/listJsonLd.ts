import {pipe} from "rxjs";
import {mergeMap} from "rxjs/operators";
import {SAMBAL_INTERNAL} from "../constants";

export function listJsonLd() {
    return pipe(
        mergeMap(async d => {
            if (d[SAMBAL_INTERNAL].jsonld && d[SAMBAL_INTERNAL].jsonld.length > 0) {
                return d[SAMBAL_INTERNAL].jsonld;
            }
            return null;
        })
    );
}