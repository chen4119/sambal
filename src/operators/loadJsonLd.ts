import {Observable, pipe} from "rxjs";
import {mergeMap} from "rxjs/operators";
import {loadContent} from "../utils";
import {hydrateJsonLd} from "sambal-jsonld";

export function loadJsonLd(hydrateOptions?: {
    fetcher?: (url: string) => Promise<any>,
    context?: any
}) {
    return pipe<Observable<string>, Observable<any>>(
        mergeMap(async src => {
            const content = await loadContent(src);
            if (hydrateOptions) {
                return await hydrateJsonLd(content, hydrateOptions.fetcher, hydrateOptions.context);
            }
            return content;
        })
    );
}