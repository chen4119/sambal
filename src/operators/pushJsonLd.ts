import {pipe, Observable} from "rxjs";
import {map} from "rxjs/operators";
import {SambalData, SAMBAL_INTERNAL} from "../constants";

export function pushJsonLd(transformer: (props: any) => any) {
    return pipe<Observable<any>, Observable<SambalData>>(
        map((data) => {
            if (!data[SAMBAL_INTERNAL]) {
                data[SAMBAL_INTERNAL] = {};
            }
            if (!data[SAMBAL_INTERNAL].jsonld) {
                data[SAMBAL_INTERNAL].jsonld = [];
            }
            const jsonld = transformer(data);
            if (jsonld) {
                data[SAMBAL_INTERNAL].jsonld.push(jsonld);
            }
            return data;
        })
    );
}
