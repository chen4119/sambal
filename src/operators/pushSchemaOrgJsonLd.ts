import {pipe, Observable} from "rxjs";
import {mergeMap} from "rxjs/operators";
import {SambalData} from "../constants";

export function pushSchemaOrgJsonLd(renderer: (props: any) => Promise<any>) {
    return pipe<Observable<SambalData>, Observable<SambalData>>(
        mergeMap(async (content) => {
            if (!content.jsonld) {
                content.jsonld = [];
            }
            const schemaOrgJsonLd = await renderer(content.data);
            if (schemaOrgJsonLd) {
                content.jsonld.push(schemaOrgJsonLd);
            }
            return content;
        })
    );
}
