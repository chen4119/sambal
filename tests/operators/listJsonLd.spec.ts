import {of, from} from "rxjs";
import {loadJsonLd} from "../../src/operators/loadJsonLd";
import {listJsonLd} from "../../src/operators/listJsonLd";
import {pushJsonLd} from "../../src/operators/pushJsonLd";
import {toSchemaOrgJsonLd} from "sambal-jsonld";

describe('listJsonLd', () => {

    it('add schema.org to head', async () => {
        const result = await from(['tests/post.md'])
        .pipe(loadJsonLd())
        .pipe(pushJsonLd(d => toSchemaOrgJsonLd(d, "BlogPosting")))
        .pipe(listJsonLd())
        .toPromise();
        expect(result).toMatchSnapshot();
    });


})