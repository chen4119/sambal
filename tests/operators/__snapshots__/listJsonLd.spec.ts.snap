// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`listJsonLd add schema.org to head 1`] = `
Array [
  Object {
    "@context": "https://schema.org",
    "@type": "BlogPosting",
    "dateCreated": "2019-03-03T19:00:00.000-05:00",
    "description": "Some description",
    "headline": "Second post",
    "keywords": Array [
      "javascript",
      "rxjs",
    ],
    "text": "<p>This is my very first post1</p>
<pre><code class=\\"language-ShellSession\\">$ npm install --save-dev sambal-cli</code></pre>
<p>Sambal has <a href=\\"https://webpack.org\\">webpack</a>, <a href=\\"https://webpack.org\\">webpack</a>, <a href=\\"https://webpack.org\\">webpack</a> to do with something else</p>
<pre><code class=\\"language-js\\">import {pipe, from, of} from &quot;rxjs&quot;;
import {mergeMap, map, groupBy, reduce, toArray} from &quot;rxjs/operators&quot;;

function aggregateBy(field) {
    return pipe(
        map((data) =&gt; data[field]),
        mergeMap(fieldValue =&gt; Array.isArray(fieldValue) ? from(fieldValue) : of(fieldValue)),
        groupBy(v =&gt; v),
        mergeMap(group$ =&gt;
            group$.pipe(reduce((acc, cur) =&gt; {
                acc.count++;
                return acc;
            }, {key: group$.key, count: 0}))    
        ),
        toArray()
    );
}</code></pre>
",
  },
]
`;
